polars/polars-ops/src/series/ops/rle.rs

```diff
@@ -0,0 +1,36 @@
+use polars_core::prelude::*;
+
+pub fn rle(s: &Series) -> PolarsResult<Series> {
+    let (s1, s2) = (s.slice(0, s.len() - 1), s.slice(1, s.len()));
+    let s_neq = s1.not_equal_missing(&s2)?;
+    let n_runs = s_neq.sum().unwrap() + 1;
+    let mut lengths = Vec::with_capacity(n_runs as usize);
+    lengths.push(1);
+    let mut vals = Series::new_empty("values", s.dtype());
+    let vals = vals.extend(&s.head(Some(1)))?.extend(&s2.filter(&s_neq)?)?;
+    let mut idx = 0;
+    for v in s_neq.into_iter() {
+        if v.unwrap() {
+            idx += 1;
+            lengths.push(1);
+        } else {
+            lengths[idx] += 1;
+        }
+    }
+
+    let outvals = vec![Series::from_vec("lengths", lengths), vals.to_owned()];
+    Ok(StructChunked::new("rle", &outvals)?.into_series())
+}
+
+pub fn rle_id(s: &Series) -> PolarsResult<Series> {
+    let (s1, s2) = (s.slice(0, s.len() - 1), s.slice(1, s.len()));
+    let s_neq = s1.not_equal_missing(&s2)?;
+
+    let mut out = Vec::with_capacity(s.len());
+    out.push(0); // Run numbers start at zero
+    s_neq
+        .into_iter()
+        .enumerate()
+        .for_each(|(i, v)| out.push(out[i] + v.unwrap() as u32));
```

Comment:
Can we keep a `latest_value` here? `out[i]` requires to keep a register for `i` and we do a bound check on `out`. The bound check may be elided by the compiler, but if we write it differently we are sure that there is no bound check.

A second optimization is that we can iterate over the arrays in `s_neq` by calling `downcast_iter`. Then we gen `BooleanArray`. From those we can directly use the `values_iter`. This saves a few branches in the iterator and the `unwrap` itself.

Reply:
OK, I tried something like that. It now dumps the inequality checks directly into the vector and does a cumsum in place.
It always feels a little awkward to iterate over each chunk but I guess that's what's really underneath so that's what has to be done.

Reply:
> It always feels a little awkward to iterate over each chunk but I guess that's what's really underneath so that's what has to be done.

Yes, then the code is closer to what we have in memory and has to go through less abstractions between the `get_value` from here and `write_value` there.